### ISSUE TITLE
Constrained decoding for valid tool-call JSON

### ISSUE THREAD

**Issue by NP-compete** [NP-compete/smithery#15]
## Problem

Even after fine-tuning, small models occasionally produce malformed tool-call JSON — wrong field names, invalid argument types, hallucinated tool names. Constrained decoding can eliminate these errors by ensuring the model's output is always valid JSON matching the tool schemas.

This is a **free accuracy boost of 5-10%** that no other tool-calling toolkit offers out of the box.

### Grammar-based Constrained Decoding

At inference time, restrict the model's token sampling to only produce tokens that lead to valid tool-call JSON.

**Level 1: JSON validity**
- Output must be syntactically valid JSON
- Use a JSON grammar (BNF / GBNF format for llama.cpp)

**Level 2: Schema compliance**
- Tool name must be one of the defined tools
- Arguments must match the tool's parameter schema
- Required parameters must be present
- Enum values must be from the allowed set

### Implementation Options

| Approach | Pros | Cons |
|---|---|---|
| **llama.cpp GBNF grammar** | Native support, no extra deps | Only works with GGUF/llama.cpp |
| **Outlines library** | Works with any HF model, powerful | Extra dependency, some latency |
| **Custom logit processor** | Full control, works with vLLM | Most complex to implement |

Recommended: support all three, with Outlines as the default for HF models and GBNF for GGUF.

### Schema → Grammar Compiler

Build a compiler that takes a set of \`ToolDefinition\` objects and produces:

1. A JSON schema that matches valid tool calls for these specific tools
2. An Outlines regex/grammar for use during inference
3. A GBNF grammar file for llama.cpp

\`\`\`python
from smithery.serve import ConstrainedDecoder

decoder = ConstrainedDecoder(tools=tools)
# For HF models (Outlines):
output = decoder.generate(model, tokenizer, prompt)
# For GGUF (llama.cpp):
grammar = decoder.to_gbnf()
\`\`\`

## Acceptance Criteria

- [ ] Schema → Outlines grammar compiler
- [ ] Schema → GBNF grammar compiler
- [ ] \`ConstrainedDecoder\` class for HF model inference
- [ ] GBNF file generation for llama.cpp / Ollama
- [ ] Benchmark: accuracy with vs without constrained decoding
- [ ] Latency impact measured (<10% overhead target)
- [ ] Works with all supported base models
- [ ] Unit tests for grammar generation from tool schemas

## Dependencies

- Depends on #9 (evaluation — to measure the accuracy boost)
- Optional dependency: \`outlines\` library

## References

- [Outlines](https://github.com/outlines-dev/outlines) — Structured generation for LLMs
- [llama.cpp GBNF](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md) — Grammar-based sampling
- [Guidance](https://github.com/guidance-ai/guidance) — Alternative structured generation

## Estimated Effort

Medium-Large (5-7 days). The grammar compiler is the hard part — handling nested objects, arrays, unions, and optionals in a BNF grammar.
